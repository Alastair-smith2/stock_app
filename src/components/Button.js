import React, { Component } from "react";
import { View, Text, TouchableHighlight, StyleSheet } from "react-native";

export default class Button extends Component {
  constructor(props) {
    super(props);

    this.onPress = this.onPress.bind(this);
  }
  onPress = () => {
    this.props.buttonPress();
  };

  render() {
    return (
      <TouchableHighlight
        onPress={this.onPress}
        style={[styles.container, this.props.buttonStyle]}
      >
        <Text style={styles.label}>{this.props.title}</Text>
      </TouchableHighlight>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1
  },
  label: {
    fontSize: 20,
    color: "#fff"
  }
});
