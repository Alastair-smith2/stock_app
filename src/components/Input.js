import React, { Component } from "react";
import { View, Text, TextInput, StyleSheet } from "react-native";

export default class Input extends Component {
  onTextChange = props => {
    this.props.onTextInput(this.props.type, props);
  };

  render() {
    return (
      <View style={styles.container}>
        <Text style={styles.label}>{this.props.label}</Text>
        <TextInput
          value={this.props.value}
          placeholder={`Please enter you ${this.props.type}`}
          onChangeText={text => this.onTextChange(text)}
          type={this.props.type}
          underlineColorAndroid={"transparent"}
          autoCapitalize={"none"}
          style={styles.textInput}
        />
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: "center",
    alignContent: "center"
  },
  label: {
    fontSize: 16,
    fontFamily: "Futura"
  },
  textInput: {
    flex: 1.5,
    height: 40,
    borderColor: "#BBB",
    borderWidth: 1,
    padding: 10
  }
});
