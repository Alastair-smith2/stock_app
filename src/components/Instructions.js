import React from "react";
import { View, Text, StyleSheet, Platform } from "react-native";

const Instructions = ({ instructionContainer, label }) => {
  return (
    <View style={instructionContainer}>
      <Text style={styles.instruction}>{label}</Text>
    </View>
  );
};

const styles = StyleSheet.create({
  instruction: {
    fontSize: 18,
    textAlign: "center",
    fontFamily: "Futura",
    ...Platform.select({
      ios: {
        color: "black"
      },
      android: {
        color: "#212121"
      }
    })
  }
});

export default Instructions;
