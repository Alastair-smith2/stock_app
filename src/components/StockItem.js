import React from "react";
import { View, Text, StyleSheet, Platform } from "react-native";

const StockItem = ({ item }) => {
  return (
    <View style={styles.container}>
      <View style={styles.date}>
        <Text style={styles.text}>Date - {item.date}</Text>
      </View>
      <View style={styles.padding} />
      <View style={styles.high}>
        <Text style={styles.text}>High- {item.data["2. high"]}</Text>
      </View>
      <View style={styles.padding} />
      <View style={styles.low}>
        <Text style={styles.text}>Low - {item.data["3. low"]}</Text>
      </View>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 0.2,
    flexDirection: "row",
    marginBottom: 10,
    justifyContent: "center"
  },
  date: {
    flex: 0.25
  },
  padding: {
    flex: 0.1
  },
  high: {
    flex: 0.25
  },
  low: {
    flex: 0.25
  },
  text: {
    textAlign: "center",
    ...Platform.select({
      ios: {
        color: "black"
      },
      android: {
        color: "#212121"
      }
    })
  }
});
export default StockItem;
