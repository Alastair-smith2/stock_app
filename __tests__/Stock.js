import React from "react";
import { shallow } from "enzyme";
import { Alert } from "react-native";
import {
  Stock,
  mapDispatchToProps,
  mapStateToProps
} from "../src/screens/Stock.js";
import Button from "../src/components/Button.js";
import Input from "../src/components/Input.js";
import Instructions from "../src/components/Instructions.js";
import StockItem from "../src/components/StockItem.js";

let mockData = {
  "Meta Data": {
    "1. Information": "Intraday (1min) prices and volumes",
    "2. Symbol": "MSFT",
    "3. Last Refreshed": "2018-03-15 16:00:00",
    "4. Interval": "1min",
    "5. Output Size": "Compact",
    "6. Time Zone": "US/Eastern"
  },
  "Monthly Time Series": {
    "2018-03-14 16:00:00": {
      "1. open": "94.2550",
      "2. high": "94.3300",
      "3. low": "94.1700",
      "4. close": "94.1800",
      "5. volume": "2415049"
    },
    "2018-03-13 15:59:00": {
      "1. open": "94.2700",
      "2. high": "94.3000",
      "3. low": "94.2100",
      "4. close": "94.2550",
      "5. volume": "217935"
    },
    "2018-03-12 15:58:00": {
      "1. open": "94.2400",
      "2. high": "94.3000",
      "3. low": "94.2300",
      "4. close": "94.2700",
      "5. volume": "143333"
    },
    "2018-03-11 15:57:00": {
      "1. open": "94.2050",
      "2. high": "94.2500",
      "3. low": "94.1900",
      "4. close": "94.2350",
      "5. volume": "277754"
    },
    "2018-03-10 15:59:00": {
      "1. open": "94.2700",
      "2. high": "94.3000",
      "3. low": "94.2100",
      "4. close": "94.2550",
      "5. volume": "217935"
    },
    "2018-03-09 15:58:00": {
      "1. open": "94.2400",
      "2. high": "94.3000",
      "3. low": "94.2300",
      "4. close": "94.2700",
      "5. volume": "143333"
    },
    "2018-03-08 15:57:00": {
      "1. open": "94.2050",
      "2. high": "94.2500",
      "3. low": "94.1900",
      "4. close": "94.2350",
      "5. volume": "277754"
    },
    "2018-03-05 15:56:00": {
      "1. open": "94.1400",
      "2. high": "94.2100",
      "3. low": "94.1300",
      "4. close": "94.2050",
      "5. volume": "381419"
    }
  }
};
let formattedData = Object.keys(mockData["Monthly Time Series"]).map(e => {
  return { date: e, data: mockData["Monthly Time Series"][e] };
});

describe("The Stock screen", () => {
  let wrapper;
  jest.mock("Alert", () => {
    return {
      alert: jest.fn()
    };
  });

  let navigation = {
    navigate: jest.fn()
  };
  const spy = jest.spyOn(Alert, "alert");
  let mockFetchStock = jest.fn();

  let appData = {
    authenticated: false,
    stockData: [],
    stockCode: "",
    currencyData: null,
    dataFetched: false,
    isFetching: false,
    error: false
  };
  let mockInitialFetchData = jest.fn();

  let fetchingData = {
    authenticated: false,
    stockData: [],
    stockCode: "",
    currencyData: null,
    dataFetched: false,
    isFetching: true,
    error: false
  };

  let errorData = {
    authenticated: false,
    stockData: [],
    stockCode: "",
    currencyData: null,
    dataFetched: false,
    isFetching: false,
    error: true
  };

  let mockAppData = {
    authenticated: false,
    stockData: formattedData,
    stockCode: "",
    currencyData: null,
    dataFetched: false,
    isFetching: false,
    error: false
  };

  beforeEach(() => {
    wrapper = shallow(
      <Stock
        appData={appData}
        navigation={navigation}
        initialFetchData={mockInitialFetchData}
        fetchStock={mockFetchStock}
      />
    );
    navigation.navigate.mockReset();
    mockInitialFetchData.mockReset();
    mockFetchStock.mockReset();
    wrapper.setProps({ appData: mockAppData });
    wrapper.setState({ readMore: true });
  });
  describe("It should render", () => {
    it("Should render", () => {
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe("Features", () => {
    describe("When there is data", () => {
      it("Should have a FlatList", () => {
        wrapper.setProps({ appData: mockAppData });
        expect(wrapper.find("FlatList")).toHaveLength(1);
      });

      it("Should have a FlatList extractor function", () => {
        let testItem = {
          date: "2018-03-05"
        };
        expect(wrapper.instance().keyExtractor(testItem)).toEqual(
          testItem.date
        );
      });

      it("Should have a FlatList renderItem function", () => {
        let item = {
          data: "Test"
        };
        expect(wrapper.instance().renderItem({ item })).toEqual(
          <StockItem item={item} />
        );
      });

      it("Should have a function that displays an arrow when not at the bottom", () => {
        // Length of data
        let insufficientData = {
          viewableItems: []
        };
        wrapper.instance().handleListChange(insufficientData);
        expect(wrapper.state("readMore")).toEqual(true);
        let dataBottom = {
          viewableItems: [
            { key: "2018-03-14 16:00:00" },
            { key: "2018-03-05 15:59:00" },
            { key: "2018-03-14 15:58:00" },
            { key: "2018-03-05 16:00:00" },
            { key: "2018-03-14 15:59:00" },
            { key: "2018-03-05 15:58:00" },
            { key: "2018-03-14 15:57:00" },
            { key: "2018-03-05 15:56:00" }
          ]
        };
        wrapper.instance().handleListChange(dataBottom);
        expect(wrapper.state("readMore")).toEqual(false);
        let dataMore = {
          viewableItems: [
            { key: "2018-03-14 16:00:00" },
            { key: "2018-03-05 15:59:00" },
            { key: "2018-03-14 15:58:00" },
            { key: "2018-03-05 16:00:00" },
            { key: "2018-03-14 15:59:00" },
            { key: "2018-03-05 15:58:00" },
            { key: "2018-03-14 15:57:00" },
            { key: "2018-03-05 15:57:00" }
          ]
        };
        wrapper.instance().handleListChange(dataMore);
        expect(wrapper.state("readMore")).toEqual(true);
      });

      it("Should have a function that displays or hides 'Read More' depending on distance from the end of the FlatList", () => {
        let tooFar = {
          distanceFromEnd: -10
        };
        wrapper.instance().endReached(tooFar);
        expect(wrapper.state("readMore")).toEqual(true);
        let closeData = {
          distanceFromEnd: 10
        };
        wrapper.instance().endReached(closeData);
        expect(wrapper.state("readMore")).toEqual(false);
      });

      it("Has a handle search function", () => {
        wrapper.instance().handleSearch();
        expect(Alert.alert.mock.calls.length).toEqual(1);
        wrapper.setState({ search: "MSFT" });
        wrapper.instance().handleSearch();
        expect(mockFetchStock.mock.calls.length).toEqual(1);
        expect(mockFetchStock.mock.calls[0][0]).toEqual("MSFT");
      });
    });

    it("Should have a loading screen", () => {
      wrapper.setProps({ appData: fetchingData });
      expect(wrapper.find("Loading")).toHaveLength(1);
    });

    describe("When there is an error", () => {
      it("Should have a Search component", () => {
        wrapper.setProps({ appData: errorData });
        expect(wrapper.find("Error")).toHaveLength(1);
      });

      it("Should be able to start over", () => {
        wrapper.setProps({ appData: errorData });
        wrapper
          .find("Error")
          .props()
          .startAgain(),
          expect(mockInitialFetchData.mock.calls.length).toEqual(1);
      });
    });

    it("Should have an instruction component", () => {
      expect(wrapper.find(Instructions)).toHaveLength(1);
    });

    it("Should an initial state of an empty search text and readMore to be true", () => {
      expect(wrapper.state("search")).toEqual("");
      expect(wrapper.state("readMore")).toEqual(true);
    });

    it("Should have a function that handles the textChange", () => {
      wrapper.instance().handleTextChange("search", "Alastair");
      expect(wrapper.state("search")).toEqual("Alastair");
    });

    describe("Map Dispatch To Props", () => {
      it("Should call the initial fetch data action", () => {
        const dispatchSpy = jest.fn();
        const { initialFetchData } = mapDispatchToProps(dispatchSpy);
        initialFetchData();
        expect(dispatchSpy.mock.calls.length).toEqual(1);
      });

      it("Should call a fetch stock action", () => {
        const dispatchSpy = jest.fn();
        const { fetchStock } = mapDispatchToProps(dispatchSpy);
        fetchStock();
        expect(dispatchSpy.mock.calls.length).toEqual(1);
      });

      it("Should return the same state as props via mapStateToProps", () => {
        let initialData = {
          appData: 1
        };
        expect(mapStateToProps(initialData).appData).toEqual(
          initialData.appData
        );
      });
    });
  });
});
