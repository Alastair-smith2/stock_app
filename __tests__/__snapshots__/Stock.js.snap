// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Input component It should render Should render 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Stock />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
      >
        I will be the stock screen
      </Text>,
      "style": Object {
        "backgroundColor": "#fff",
        "flex": 1,
      },
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "accessible": true,
        "allowFontScaling": true,
        "children": "I will be the stock screen",
        "ellipsizeMode": "tail",
      },
      "ref": null,
      "rendered": "I will be the stock screen",
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
        >
          I will be the stock screen
        </Text>,
        "style": Object {
          "backgroundColor": "#fff",
          "flex": 1,
        },
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "accessible": true,
          "allowFontScaling": true,
          "children": "I will be the stock screen",
          "ellipsizeMode": "tail",
        },
        "ref": null,
        "rendered": "I will be the stock screen",
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
